sql/core/src/main/scala/org/apache/spark/sql/catalyst/util/ExpressionSQLBuilder.scala

```diff
@@ -0,0 +1,56 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.util
+
+import org.apache.spark.sql.catalyst.expressions.{CaseWhen, Expression, Literal}
+import org.apache.spark.sql.connector.expressions.{FieldReference, GeneralSQLExpression, LiteralValue}
+import org.apache.spark.sql.execution.datasources.PushableColumnWithoutNestedColumn
+import org.apache.spark.sql.execution.datasources.v2.DataSourceV2Strategy
+
+/**
+ * The builder to build {@link GeneralSQLExpression}.
+ */
+class ExpressionSQLBuilder(e: Expression) {
+
+  def build(): Option[GeneralSQLExpression] = e match {
+    case CaseWhen(branches, elseValue) =>
+      val newBranches = branches.collect {
+        case (predicate: Expression, Literal(value, dataType)) =>
```

Comment:
We should implement this SQL string builder in a more flexible way. Something like
```
def build = generateSQL(e)
private def generateSQL(expr: Expression): Option[String] = expr match {
  case l: Literal => Some(...)
  case PushableColumnWithoutNestedColumn(name) => Some(name)
  // other basic expressions, like predicates?
  case CaseWhen(branches, elseValue) =>
    val conditionsSQL = branches.map(_._1).flatMap(generateSQL)
    val valuesSQL = branches.map(_._2).flatMap(generateSQL)
    ...
}
```